### ISSUE TITLE
Logging trough Attribute

### ISSUE THREAD

**Issue by DMedo** [pieceofsummer/Hangfire.Console#28]
is it possible to use this logger with StateContext instead of PerformContext?

**Comment by pieceofsummer** [comment]
Leaving aside the fact that `StateContext` Is an obsolete class (which itself is a good enough reason not to bother with it), both its subclasses `ElectStateContext` and `ApplyStateContext` are contexts for state transitions. Being a state rendered, Console requires a job state to be bound to. Since there's no fixed state in the middle of state transition, I don't see how you expect it to work.

**Comment by DMedo** [comment]
`using System;
using System.Collections.Generic;
using Hangfire;
using Hangfire.Client;
using Hangfire.Common;
using Hangfire.Console;
using Hangfire.Server;
using Hangfire.States;
using Hangfire.Storage;

namespace Test
{
    public class DisableMultipleQueuedItemsAttribute : JobFilterAttribute, IClientFilter, IServerFilter, IElectStateFilter
    {
        private static readonly TimeSpan LockTimeout = TimeSpan.FromSeconds(10);
        private static readonly List<string> ReleaseStates = new List<string> { DeletedState.StateName, FailedState.StateName };

        private bool LockSet;

        public void OnCreating(CreatingContext filterContext)
        {
        }

        public void OnCreated(CreatedContext filterContext)
        {
        }

        public void OnPerforming(PerformingContext filterContext)
        {

            if (!AddFingerprintIfNotExists(filterContext.Connection, filterContext.BackgroundJob))
            {
                filterContext.Canceled = true;
                filterContext.WriteLine("OnPerforming: Fingerprint found, job cancelled");

                return;
            }

            filterContext.WriteLine("OnPerforming: Fingerprint created");
        }

        public void OnPerformed(PerformedContext filterContext)
        {
            // If job finished with success we remove lock.
            if (filterContext.Exception == null || filterContext.ExceptionHandled)
            {
                RemoveFingerprint(filterContext.Connection, filterContext.BackgroundJob);

                filterContext.WriteLine("OnPerformed: Fingerprint removed");
            }
        }

        public void OnStateElection(ElectStateContext context)
        {
            // If job is going to release state and it isn't in failed state we release the lock.
            // if job is already in failed state we don't release lock because lock can be from a running job, failed job already had it's lock removed.
            if (ReleaseStates.Contains(context.CandidateState.Name) && context.CurrentState != FailedState.StateName)
            {
                RemoveFingerprint(context.Connection, context.BackgroundJob);
            }
        }

        private bool AddFingerprintIfNotExists(IStorageConnection connection, Job job)
        {
            if (LockSet)
            {
                return LockSet;
            }

            using (connection.AcquireDistributedLock(GetFingerprintLockKey(job), LockTimeout))
            {
                var fingerprint = connection.GetAllEntriesFromHash(GetFingerprintKey(job));

                if (fingerprint != null && fingerprint.ContainsKey("Timestamp"))
                {
                    // Actual fingerprint found, returning.
                    return false;
                }

                // Fingerprint does not exist, it is invalid (no `Timestamp` key),
                // or it is not actual (timeout expired).
                connection.SetRangeInHash(
                    GetFingerprintKey(job),
                    new Dictionary<string, string>
                    {
                        { "Timestamp", DateTimeOffset.UtcNow.ToString("o") }
                    });

                LockSet = true;

                return true;
            }
        }

        private bool AddFingerprintIfNotExists(IStorageConnection connection, BackgroundJob job)
        {
            return AddFingerprintIfNotExists(connection, job.Job);
        }

        private string GetFingerprint(Job job)
        {
            string parameters = string.Empty;
            if (job.Args != null)
            {
                parameters = string.Join(".", job.Args);
            }

            if (job.Type == null || job.Method == null)
            {
                return string.Empty;
            }

            var fingerprint = $"{job.Type.Name}.{job.Method.Name}.{parameters}";

            return fingerprint;
        }

        private string GetFingerprintKey(Job job)
        {
            return $"Fingerprint:{GetFingerprint(job)}";
        }

        private string GetFingerprintLockKey(Job job)
        {
            return $"{GetFingerprintKey(job)}:lock";
        }

        private void RemoveFingerprint(IStorageConnection connection, BackgroundJob job)
        {
            RemoveFingerprint(connection, job.Job);
        }

        private void RemoveFingerprint(IStorageConnection connection, Job job)
        {
            using (connection.AcquireDistributedLock(GetFingerprintLockKey(job), LockTimeout))
            {
                using (var transaction = connection.CreateWriteTransaction())
                {
                    transaction.RemoveHash(GetFingerprintKey(job));
                    transaction.Commit();
                }
            }
        }
    }
}`

I have attribute like this, I'd like to log from all methods if possible. Not only OnPerforming and OnPerformed. Is that possible?

**Comment by pieceofsummer** [comment]
Nope, that's not possible. As I've said, Console is a state renderer, so unless it is `Processing` state, it has nowhere to render to. 

It is not even advised to use it from OnPerforming/OnPerformed handlers, because it might or might not work, depending on filters' registration order.

It is not a debug logger, to be available to log from everywhere. It is designed to track job progress, not the correctness of filters implementation.

**Comment by DMedo** [comment]
That's what I thought, do you know of any other library that would allow something like this?

**Comment by pieceofsummer** [comment]
I believe, the built-in logging library (LibLog) would work.

**Comment by DMedo** [comment]
It could be shown on dashboard?

**Comment by pieceofsummer** [comment]
No, it couldn't. Dashboard is not designed for debugging purposes, either.
